cpp/src/arrow/compute/kernels/vector_nested.cc

```diff
@@ -77,24 +146,53 @@ const FunctionDoc list_parent_indices_doc(
      "is emitted."),
     {"lists"});
 
+class ListParentIndicesFunction : public MetaFunction {
+ public:
+  ListParentIndicesFunction()
+      : MetaFunction("list_parent_indices", Arity::Unary(), &list_parent_indices_doc) {}
+
+  Result<Datum> ExecuteImpl(const std::vector<Datum>& args,
+                            const FunctionOptions* options,
+                            ExecContext* ctx) const override {
+    KernelContext kernel_ctx(ctx);
+    switch (args[0].kind()) {
+      case Datum::ARRAY:
+        return ListParentIndicesArray::Exec(&kernel_ctx, args[0].array());
+      case Datum::CHUNKED_ARRAY: {
+        const auto& input = args[0].chunked_array();
+        ARROW_ASSIGN_OR_RAISE(auto out_ty, ListParentIndicesType(*input->type()));
+
+        int64_t base_output_offset = 0;
+        ArrayVector out_chunks;
+        for (const auto& chunk : input->chunks()) {
+          ARROW_ASSIGN_OR_RAISE(auto out_chunk,
+                                ListParentIndicesArray::Exec(&kernel_ctx, chunk->data(),
+                                                             base_output_offset));
+          out_chunks.push_back(MakeArray(std::move(out_chunk)));
+          base_output_offset += chunk->length();
+        }
+        return std::make_shared<ChunkedArray>(std::move(out_chunks), std::move(out_ty));
+      }
+      default:
+        return Status::NotImplemented(
+            "Unsupported input type for function 'list_parent_indices': ",
+            args[0].ToString());
+    }
+  }
+};
+
 }  // namespace
 
 void RegisterVectorNested(FunctionRegistry* registry) {
   auto flatten =
       std::make_shared<VectorFunction>("list_flatten", Arity::Unary(), &list_flatten_doc);
-  DCHECK_OK(flatten->AddKernel({InputType::Array(Type::LIST)}, OutputType(ValuesType),
+  DCHECK_OK(flatten->AddKernel({InputType::Array(Type::LIST)}, OutputType(ListValuesType),
                                ListFlatten<ListType>));
   DCHECK_OK(flatten->AddKernel({InputType::Array(Type::LARGE_LIST)},
-                               OutputType(ValuesType), ListFlatten<LargeListType>));
+                               OutputType(ListValuesType), ListFlatten<LargeListType>));
   DCHECK_OK(registry->AddFunction(std::move(flatten)));
 
-  auto list_parent_indices = std::make_shared<VectorFunction>(
-      "list_parent_indices", Arity::Unary(), &list_parent_indices_doc);
-  DCHECK_OK(list_parent_indices->AddKernel({InputType::Array(Type::LIST)}, int32(),
-                                           ListParentIndices<ListType>));
-  DCHECK_OK(list_parent_indices->AddKernel({InputType::Array(Type::LARGE_LIST)}, int64(),
-                                           ListParentIndices<LargeListType>));
```

Comment:
The old ListParentIndices implementation can be removed, no?

Reply:
Hmm, definitely!